Training/virtual_functions.cpp

```diff
@@ -6,89 +6,140 @@ using namespace std;
-        }
-        cout << " " << marks_sum << " " << m_cur_id << endl;
-    }
+	Student ();
+	static Person * make ();
+	void getdata () override;
+	void putdata () override;
 };
 
-int Student::s_next_id = 1;
 
-int hackerrank::virtual_functions()
+int hackerrank::virtual_functions ()
 {
-    // START HACKERRANK BOILERPLATE
-    int n, val;
-    cin >> n; //The number of objects that is going to be created.
-    Person *per[1000];
+	// START HACKERRANK BOILERPLATE
+	int n, val;
+	cin >> n; //The number of objects that is going to be created.
+	Person * per[1000];
+	memset(per, 0, sizeof(Person*)); ///Es-Note: Initialize `per` array to nullptrs for safety.
+
+	for (int i = 0; i < n; i++) {
+
+		cin >> val;
+		if (val == 1) {
+			// If val is 1 current object is of type Professor
+			per[i] = new Professor(); ///Es-Note: Use function-call style syntax for constructors.
+
+		}
+		else { ///Es-Note: match brace style w/the if statement-- either both or neither have it.
+			per[i] = new Student(); // Else the current object is of type Student
+		}
+		per[i]->getdata(); // Get the data from the user.
+
+	}
+
+	for (int i = 0; i < n; i++)
+		per[i]->putdata(); // Print the required output for each object.
+
+	return 0;
+	// END HACKERRANK BOILERPLATE
+}
+
+
+///////////////////
+// Professor Impl
 
-    for (int i = 0;i < n;i++) {
+int Professor::s_next_id = 1;
 
-        cin >> val;
-        if (val == 1) {
-            // If val is 1 current object is of type Professor
-            per[i] = new Professor;
+inline Professor::Professor () {
+	m_cur_id = s_next_id++;
+}
+inline void Professor::getdata () {
+	cin >> m_name >> m_age >> m_publications;
+}
+inline void Professor::putdata () {
+	cout << m_name << " " << m_age << " " << m_publications << " " << m_cur_id << endl;
+}
 
-        }
-        else per[i] = new Student; // Else the current object is of type Student
+inline Person * Professor::make () {
+	auto ret = new Professor();
+	ret->m_cur_id = s_next_id;
+	return ret;
+}
+
+///////////////////
+// Student Impl
 
-        per[i]->getdata(); // Get the data from the user.
+/// Es-Note: Static variables/objects are always default initialized to 0 or the  default constructor.
+/// In this case, you could work with this assumption and just start counting from 0.
+// int Student::s_next_id = 1;
 
-    }
+inline Student::Student () {
+	///Es-Note: Constructors get called incidentally/accidentally all the time.
+	/// It's a good idea not to perform static incrementation using a constructor,
+	/// but instead provide a factory builder. That way this object will behave
+	/// when used in data structures that need to copy stuff around.
+	/// IE. Just because the object was constructed, does not mean that it
+	/// represents new data.
+	m_cur_id = s_next_id++;
+}
+
+
+inline Person * Student::make () {
+	/// Es-Note: See 'Student::Student ()' above.
+	auto ret = new Student();
+	ret->m_cur_id = ++s_next_id; // Assuming s_next_id starts at 0.
```

Comment:
I thought about doing this but it seemed a little wrong. Either the variable should be named `s_last_id` or it should actually hold what the next ID is. Unfortunately, `s_last_id` is kind of also a lie because it begins at 0 and the ID 0 is never used.

Reply:
Was the var name part of the boiler? s_id_counter would be fine.

Reply:
The var name wasn't part of the given code. `s_id_counter` is fine.